### PR TITLE
Migrate crumbs to a function

### PR DESCRIPTION
**Pull request by Quy**
Create a breadcrumb function to generate the breadcrumb markup.

Refs #1128

### CODE REVIEW

File: misc.php [hunk 1]
```diff
@@ -275,6 +275,12 @@
 	if ($pun_config['o_censoring'] == '1')
 		$cur_post['subject'] = censor_words($cur_post['subject']);
 
+	$crumbs = generate_crumbs(array(
+		array($lang_common['Index'], 'index.php'),
+		array(pun_htmlspecialchars($cur_post['forum_name']), 'viewforum.php?id='.$cur_post['fid']),
+		array(pun_htmlspecialchars($cur_post['subject']), 'viewtopic.php?pid='.$post_id.'#p'.$post_id),
```
Comment (franzliedke) [inline misc.php]: What about moving the `pun_htmlspecialchars` calls to the function, too?

Comment (Quy) [reply misc.php]: I did it like the following where some entries require it and others don't like language strings.

`$page_title = array(pun_htmlspecialchars($pun_config['o_board_title']), $lang_common['User list']);`

Comment (franzliedke) [reply misc.php]: Yeah, I saw that. But it won't be a problem with language strings - and I like APIs that do the safe thing for you automatically, which means it cannot be forgotten.,

Comment (Quy) [reply misc.php]: Sounds good! Will do.

Comment (Quy) [reply misc.php]: Done.

File: include/functions.php [hunk 2]
```diff
@@ -574,6 +574,27 @@ function generate_page_title($page_title, $p = null)
...
+function generate_crumbs($crumbs)
+{
+	$last_key = end(array_keys($crumbs));
+
+	$crumbs_markup = '<ol class="crumbs">';
+	foreach ($crumbs as $idx => $crumb)
+	{
+		$crumbs_markup .= '<li>'.($idx ? '<span>»&#160;</span>' : '');
+		$item = is_array($crumb) ? '<a href="'.$crumb[1].'">'.pun_htmlspecialchars($crumb[0]).'</a>' : $crumb;
+		$crumbs_markup .= $idx == $last_key ? '<strong>'.$item.'</strong>' : $item;
```
Comment (franzliedke) [inline include/functions.php]: For consistency, I'd like to see `$crumb` (one line above) and `$item` be escaped, too. Probably not necessary for the current usages, but again safer than having to remember to add it when it becomes necessary.

File: edit.php [hunk 3]
```diff
@@ -142,7 +142,11 @@
 	}
 }
 
-
+$crumbs = generate_crumbs(array(
+	array($lang_common['Index'], 'index.php'),
+	array($cur_post['forum_name'], 'viewforum.php?id='.$cur_post['fid']),
+	array($cur_post['subject'], 'viewtopic.php?id='.$cur_post['tid']),
+	$lang_post['Edit post']));
```
Comment (franzliedke) [inline edit.php]: For all of these, please move the closing paren to an extra line, and add a trailing comma to the last array element.

File: include/functions.php (hunk 2 above)
Comment (Quy) [reply include/functions.php]: Good catch! Here is another way to do this. Do you have a preference?

```
function generate_crumbs($crumbs)
{
	$crumbs_markup = '<ol class="crumbs">';

	$num_crumbs = count($crumbs);
	for ($i = 0; $i < $num_crumbs; ++$i)
	{
		$crumbs_markup .= '<li>'.(($i) ? '<span>»&#160;</span>' : '');
		$item = (is_array($crumbs[$i])) ? '<a href="'.$crumbs[$i][1].'">'.pun_htmlspecialchars($crumbs[$i][0]).'</a>' : pun_htmlspecialchars($crumbs[$i]);
		$crumbs_markup .= ($i == $num_crumbs - 1) ? '<strong>'.$item.'</strong>' : $item;
		$crumbs_markup .= '</li>';
	}

	$crumbs_markup .= '</ol>';

	return $crumbs_markup;
}
```

Comment (franzliedke) [reply include/functions.php]: No strong preference. I find the original version a little easier to read.

You could even *assume* there are at least two elements in `$crumbs` and handle the two special cases (first and last iteration) outside of the loop...

File: misc.php (hunk 1 above)
Comment (Quy) [reply misc.php]: Aargh. It is not going to work with Search results.
```
    Index » Search » <a href="search.php?action=show_recent">Active</a>
```

File: include/functions.php (hunk 2 above)
Comment (Quy) [reply include/functions.php]: When `pun_htmlspecialchars` applied to the last `$crumb`:
```
    Index » Search » <a href="search.php?action=show_recent">Active</a>
```

Comment (franzliedke) [reply include/functions.php]: You'll have to unpack the `$crumbs_text` stuff in search.php. All that is doing is wrapping text in a link, as far as I can tell - and you can pass these two in an array to your `generate_crumbs()` function.

Comment (Quy) [reply include/functions.php]: Can you show the code for the first and last iteration?

Comment (Quy) [reply include/functions.php]: I got it to work except for the following:

Original code:
```
$crumbs_text['search_type'] = '<a href="search.php?action=search&amp;keywords='.urlencode($keywords).'&amp;author='.urlencode($author).'&amp;forums='.$search_type[2].'&amp;search_in='.$search_type[3].'&amp;sort_by='.$sort_by.'&amp;sort_dir='.$sort_dir.'&amp;show_as='.$show_as.'">'.$crumbs_text['search_type'].'</a>';
```
Revised code:
```
$crumbs_text['search_type'] = array($crumbs_text['search_type'], 'search.php?action=search&amp;keywords='.urlencode($keywords).'&amp;author='.urlencode($author).'&amp;forums='.$search_type[2].'&amp;search_in='.$search_type[3].'&amp;sort_by='.$sort_by.'&amp;sort_dir='.$sort_dir.'&amp;show_as='.$show_as);
```
Output when the keyword search is `<strong>`:
```
Index » Search » Topics with posts containing '&lt;strong&gt;'
```
Markup:
```
<ol class="crumbs"><li><a href="index.php">Index</a></li><li><span>»&#160;</span><a href="search.php">Search</a></li><li><span>»&#160;</span><strong><a href="search.php?action=search&amp;keywords=%3Cstrong%3E&amp;author=&amp;forums=&amp;search_in=0&amp;sort_by=0&amp;sort_dir=DESC&amp;show_as=topics">Topics with posts containing &#039;&amp;lt;strong&amp;gt;&#039;</a></strong></li></ol>
```

Comment (Quy) [reply include/functions.php]: We may have to revert `pun_htmlspecialchars` to have more flexibility. If yes, is there a way to revert the changes from yesterday?

Comment (franzliedke) [reply include/functions.php]: You can [undo commits with Git](https://stackoverflow.com/questions/927358/how-do-i-undo-the-most-recent-commits-in-git), and then force-push.

I can have a look again next week, will be AFK for a few days.

**Comment by Quy** [comment]
Redo of #223

**Pull request closed without merging**